[REEF-1423] Tasks are not disposed after they are closed

jwang98052 opened the pull request:
Updated TaskRunTime to ensure the Dispose is called after closing the task
Added test cases.

JIRA: [REEF-1423](https://issues.apache.org/jira/browse/REEF-1423)
This closes

afchung commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs:
```diff
@@ -173,6 +173,23 @@ public void Close(byte[] message)
             }
+            finally
+            {
+                Logger.Log(Level.Info, "TaskRuntime.Close, finally._userTask: " + _userTask);
```
You should validate that the `ITask` thread is finished before calling `Dispose`.

jwang98052 replied:
That's true. Ideally task should implement the proper close logic, that is called cooperative cancellation model  (https://msdn.microsoft.com/en-us/library/dd997364(v=vs.110).aspx). We already allow the task to add its own close handler. It is the task's responsibility to let the thread to return gracefully. 

Even if we add CancellationToken in the Call() method, it is still up to the client to respect it in the code. We have no way to control. It still falls to Cooperative cancellation model .
 
Checking if the thread is still Alive and Abort it might be a way, but Abort() is not recommended approach. 

In the current logic, at least we disposed object to release the resource. If the thread is still running, and find the object is disposed, it should throw exception and it will reach to the catch inside StartTaskOnNewThread in TaskRuntime. 
If you have any better solution, please share it.

andrey-me replied:
for consistency, in the logger.log it's probably better to use template string instead of concatenation.

jwang98052 replied:
I will remove those logs as they were used for testing.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -164,24 +164,28 @@ public byte[] Call(byte[] memento)
         {
-            var msg = Encoding.UTF8.GetString(closeEvent.Value.Value);
-            if (closeEvent.Value.IsPresent() && msg.Equals(TaskManager.CloseTaskByDriver))
+            Interlocked.Exchange(ref _shouldCloseTask, 1);
```
This logic seems the same as in `MapTaskHost`. Can we reuse the code?

jwang98052 replied:
Yes, they are the same. 
We can write a separate CloseHandler for both tasks to share, but as it uses instance variables in the tasks, put the handler inside the task class is convenient.
Writing a base class for those two? We usually write flatten task to make it easy to read. 
Do you have any suggestions?

afchung replied:
You can try to encapsulate the logic in a class.

afchung commented:
Done with a first pass.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -144,24 +144,28 @@ public byte[] Call(byte[] memento)
-                    throw new IMRUTaskSystemException(TaskManager.TaskKilledByDriver);
+                    Logger.Log(Level.Info, "The task received close event with message: {0}.", msg);
+
+                    if (Interlocked.Read(ref _isTaskStopped) == 0)
```
If the task is stuck, this value will keep as 0.

afchung replied:
No, but you set it to 1 in line 155. The code will never reach here without passing line 155.

jwang98052 replied:
Line 155 set _shouldCloseTask, here I check _isTaskStopped. Did I look at wrong file?

afchung replied:
Sorry my bad, not sure what I was looking at. Thanks!

jwang98052 commented:
@afchung Thanks for the review! I left some comments in responding to your comments. Feel free to discuss.

jwang98052 commented:
@afchung I have pushed the change. Let me know if _isTaskStopped is still an issue.

jwang98052 commented:
@afchung I have address your review comment about close handler. I also rebased the code that merged the IMRUCloseTaskTest.TestTaskCloseOnLocalRuntime I added in REEF-1408 which covers the test for the code change in this PR. Thanks!

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -133,37 +116,22 @@ public byte[] Call(byte[] memento)
+            if (_taskCloseHandler.ShouldCloseTask())
             {
-                _waitToCloseEvent.Set();
+                _taskCloseHandler.Signal();
```
You should signal regardless.

jwang98052 replied:
If ShouldCloseTask is 0, meaning the close event handler is never called, then no one is waiting.

afchung replied:
Close is asynchronous. It's possible that the close Event handler comes in a bit later and in that case it will hang.

jwang98052 replied:
If it is not waiting, call set won't harm, right?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -109,7 +92,7 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask, IObserver<IClo
         public byte[] Call(byte[] memento)
         {
-            while (Interlocked.Read(ref _shouldCloseTask) == 0)
+            while (!_taskCloseHandler.ShouldCloseTask())
```
Can you put the `break` condition check here instead of using the `break` keyword?

jwang98052 replied:
Are you talking about this line:
 while (!_taskCloseHandler.ShouldCloseTask())
There is no break keyword here.

There is a break below it:
                if (mapInput.ControlMessage == MapControlMessage.Stop)
                    {
                        break;
                    }
Do you mean here? It was original code.

afchung replied:
Yes, I think it should be moved to the top of the while loop instead of using a `break`.

jwang98052 replied:
I am not sure why he put MapInputWithControlMessage<TMapInput> mapInput = _dataAndMessageReceiver.Receive() in using, does he want to dispose it after receiving each message? With this using in the middle, I cannot put the condition
mapInput.ControlMessage == MapControlMessage.Stop in the while statement unless assign it to another variable. What do you think?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseHandler.cs:
```diff
@@ -0,0 +1,108 @@
+            Interlocked.Exchange(ref _shouldCloseTask, 1);
+            _waitToCloseEvent.Wait(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
+
+            if (closeEvent.Value.IsPresent())
```
Are we ever expecting a not-present value?

jwang98052 replied:
It is possible the close handler is called through Context.Dispose(). In that case, the value is not present. I would expect it returns without throw any exception in that case.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -145,45 +128,30 @@ public byte[] Call(byte[] memento)
+            if (_taskCloseHandler.ShouldCloseTask())
             {
-                _waitToCloseEvent.Set();
+                _taskCloseHandler.Signal();
```
signal regardless of condition.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseHandler.cs:
```diff
@@ -0,0 +1,108 @@
+
+namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
+{
+    internal sealed class TaskCloseHandler
```
Document.

afchung replied:
This is not a handler in of itself. Please rename.

jwang98052 replied:
Done

afchung commented:
Done with another pass.

jwang98052 commented:
@afchung I have addressed your review comments. Please have another look.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
+{
+    /// <summary>
+    /// This class provides a method to handler Task close event. It is called from TaskCloseEventHandler. 
```
handler -> handle

jwang98052 replied:
Done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+        private static readonly Logger Logger = Logger.GetLogger(typeof(TaskCloseCoordinator));
+
+        /// <summary>
+        /// When receiving a close event, this variable is set to 1. At the beginning of each task iteration,
```
"When receiving a close event" -> "When close event is received"

jwang98052 replied:
Done

dkm2110 commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestDisposeTasks.cs:
```diff
@@ -0,0 +1,304 @@
+            {
+                Assert.Equal(TaskId + "2", value.Id);
+
+                var failedExeption = ByteUtilities.ByteArraysToString(value.Data.Value);
```
failedExeption => failedException

jwang98052 replied:
Done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+        {
+            Interlocked.Exchange(ref _shouldCloseTask, 1);
+            _waitToCloseEvent.Wait(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
+
```
Shouldn't this wait go inside the if loop?

jwang98052 replied:
No. No matter what message is received in the close event, we do want to inform the task to close. Close can be called from other places as well. And before the task is returned, it will notify _waitToCloseEvent to continue. 
The if portion is for a specific close command from driver with message CloseTaskByDriver, we want throw exception for that particular command so that driver can reaction accordingly.

dkm2110 replied:
I see ok.

dkm2110 commented:
@jwang98052 did a pass....some very minor comments....Let me know when @afchung  signs off and I will test on cluster and merge.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs:
```diff
@@ -174,6 +174,21 @@ public void Close(byte[] message)
+                }
+                catch (Exception e)
+                {
+                    Utilities.Diagnostics.Exceptions.Caught(
```
This swallows the exception. Why is this a good idea? Also, if it is, we should document this behavior in the {{ITask}} interface documentation.

jwang98052 replied:
We have caught all the exceptions during the task closing. This is the exception during Dispose in finally block. As the task has been closed, therefore we log the error and ignore the exception.
Same code for StartTaskOnNewThread above in the same class.
It is a good point to document it. I have updated ITask comments for this.

markusweimer replied:
This is dangerous. If the finally block fails, we might leak resources. Shouldn't we fail the Evaluator in this case?

jwang98052 replied:
That's true. But if we fail Evalutor in this case, is it recoverable?

markusweimer replied:
No, and that would be correct, wouldn't it be? If we reach this state, the task's `Dispose` method threw an exception. There is no clean way to recover from that, as far as I know.

jwang98052 replied:
Changed to throw an exception.

jwang98052 commented:
@afchung @dkm2110 @markusweimer I have addressed your review comments. Please have another look and sign off if you think it is OK before Dhruv test it on cluster. Thanks!

afchung commented:
Taking a second pass.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+                        throw new IMRUTaskSystemException(TaskManager.TaskKilledByDriver);
+                    }
+                }
+            }
```
If this is not present, what does it indicate?

jwang98052 replied:
Usually when driver calls ActiveContext.Dispose() , task should have either completed or failed. ContextRunTime would log "Received a close task while there was no task running. Ignored" instead of calling task close handler. 
However, if driver calls ActiveContext.Dispose() on a running task, task close handler will be  called without any message. That is the case Event.Value.IsPresent() is false in which, we should still close the task. 
One of the test cases is to walk through this scenario.

afchung replied:
That's fair. Perhaps we should still log something here in that case?

afchung replied:
Since that behavior is not entirely expected.

jwang98052 replied:
Yes, I will add a log.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+        /// Before the task is returned, this variable is set to 1.
+        /// Close handler will check this variable to decide if it needs to throw an exception.
+        /// </summary>
+        private long _isTaskStopped = 0;
```
I don't think you actually need this. The `_waitToCloseEvent` should be enough.

jwang98052 replied:
It is needed.
In HandleEvent method, waitToCloseEvent.Wait can come out either by "set" before the task returns properly, or time out while task is hung. In the end, we need to use this flag to decide if we should throw exception. If the task was returned properly, then no need to throw exception, the driver will receive CompletedTask event. If _isTaskStopped  is not 1, meaning the task was not returned, but hung or something is wrong, we need to throw exception for driver to get FailedTask event.

afchung replied:
Wait returns a `bool` for that purpose.

jwang98052 replied:
I see your point. Then it can be leveraged.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+            Interlocked.Exchange(ref _isTaskStopped, 1);
+        }
+
+        internal void Signal()
```
Remove `_isTaskStopped` and rename this to `SignalTaskStop`.

jwang98052 replied:
I can combine SetTaskStopped and Signal as they are always called together. _isTaskStopped  is still needed because _waitToCloseEvent.Wait can get out by timeout as well, set is not the only way for it to get out.

afchung replied:
See other comment.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,113 @@
+    /// It also wraps flags to represent if the task should be closed and if the task has been stopped
+    /// so that to provide a coordination between the task and the close handler.  
+    /// </summary>
+    internal sealed class TaskCloseCoordinator
```
`ThreadSafe`.

jwang98052 replied:
Done

afchung commented:
Sorry for pointing this out a bit late, but I'm confused by what the item actually aims to do. Why are there changes in IMRU at all?

afchung commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs:
```diff
@@ -174,6 +174,21 @@ public void Close(byte[] message)
+                {
+                    if (_userTask != null)
+                    {
+                        _userTask.Dispose();
```
This logic can also get called when the Task finishes in another thread at the same time. You will need ensure that `Dispose()` is only called once, even though it is best practice for be `Dispose()` to be reentrant.

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUCloseTaskTest.cs:
```diff
@@ -52,17 +52,17 @@ public class IMRUCloseTaskTest : IMRUBrodcastReduceTestBase
-            int updateTaskMemory = 5120;
-            int numTasks = 4;
-            string testFolder = DefaultRuntimeFolder + TestId;
+            const int chunkSize = 2;
```
Changes in this file seem unnecessary.

jwang98052 replied:
I was to trigger Jankins and made some changes on purpose. The change itself is valid anyway.

afchung commented:
I'm done with another pass. I'll take a look at the tests once I understand what this PR is really trying to address.

jwang98052 commented:
@afchung The purpose of the PR is to make sure the tasks are dispose after close is called. When I was working on the test cases, I find a scenario where the message in the CloseEvent can be null, but IMRUTask Close handlers don't handle that case, it is a bug. Without fixing it, the test won't pass. Therefore I made some minor changes in the IMRUTask close handler. In your review, you pointed the two close handlers have the same code, therefore introducing a code refactor in the task close handler.

jwang98052 commented:
@afchung I have addressed your comments. Please have another look.

afchung commented:
@jwang98052 can you take a look at the AppVeyor test failure?

jwang98052 commented:
@afchung Tests failed in reading log file in AppVeyor, it happened randomly. Same issue we saw in the past. Looks like it is timing related.

afchung commented:
I'll take another look.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs:
```diff
@@ -0,0 +1,111 @@
+        internal void HandleEvent(ICloseEvent closeEvent)
+        {
+            Interlocked.Exchange(ref _shouldCloseTask, 1);
+            var taskSiganled = _waitToCloseEvent.Wait(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
```
`taskSignaled`

jwang98052 replied:
done

afchung commented:
@jwang98052 few minor comments, otherwise looks good to me.

jwang98052 commented:
@afchung I have addressed your review comments. 
@dkm2110 Would you like to test it on cluster?

jwang98052 commented:
AppVeyor passed from my server for this push:
https://ci.appveyor.com/project/jwang98052/reef

jwang98052 commented:
I have run IMRU on the cluster with the following arguments:
nodes = 1000 
memory  = 1024 MB mapper and update Evaluators
10 iterations
dim = 20000000 
chunk = 1000000

The test is successful and result looks good. 

@dkm2110 @afchung Do you still have any comments before the merge?

dkm2110 commented:
@jwang98052  there is one comment by @markusweimer  regarding swallowing the exception. @markusweimer can you look at that and see if it is still an issue. Otherwise LGTM.

jwang98052 commented:
@markusweimer I have addressed your comment. Please have another look.

markusweimer commented:
LGTM

jwang98052 commented:
@afchung Do you still have any comment?

jwang98052 commented:
@afchung Do you still have any comment before the merge?

afchung commented:
No, looks good to me, as mentioned above.

jwang98052 commented:
AppVeyor passed in my server
https://ci.appveyor.com/project/jwang98052/reef/build/37-REEF-1423

tcNickolas commented:
All reviewers signed off, so I'll jump in and merge :-) Tests pass on my local machine as well
